source/opt/constant_value.h

```diff
@@ -0,0 +1,304 @@
+
+  virtual uint32_t NumComponents() const {
+    return static_cast<uint32_t>(components_.size());
+  }
+  virtual const Const* GetComponent(uint32_t index) const {
+    assert(static_cast<size_t>(index) < components_.size() &&
+           "GetComponent() index out of bound.");
+    return components_[index];
+  }
+
+ protected:
+  CompositeConst(const analysis::Type* ty) : Const(ty), components_() {}
+  CompositeConst(const analysis::Type* ty,
+                 const std::vector<const Const*>& components)
+      : Const(ty), components_(components) {}
+  CompositeConst(const analysis::Type* ty,
+                 std::vector<const Const*>&& components)
+      : Const(ty), components_(std::move(components)) {}
+  std::vector<const Const*> components_;
+};
+
+// Struct type constant.
+class StructConst : public CompositeConst {
+ public:
+  StructConst(const analysis::Struct* ty) : CompositeConst(ty) {}
+  StructConst(const analysis::Struct* ty,
+              const std::vector<const Const*>& components)
+      : CompositeConst(ty, components) {}
+  StructConst(const analysis::Struct* ty,
+              std::vector<const Const*>&& components)
+      : CompositeConst(ty, std::move(components)) {}
+
+  StructConst* AsStructConst() override { return this; }
+  const StructConst* AsStructConst() const override { return this; }
+
+  std::unique_ptr<StructConst> CopyStructConst() const {
+    return MakeUnique<StructConst>(type_->AsStruct(), components_);
+  }
+  std::unique_ptr<Const> Copy() const override {
+    return std::unique_ptr<Const>(CopyStructConst().release());
+  }
+};
+
+// Vector type constant.
+class VectorConst : public CompositeConst {
+ public:
+  VectorConst(const analysis::Vector* ty)
+      : CompositeConst(ty), component_type_(ty->element_type()) {}
+  VectorConst(const analysis::Vector* ty,
+              const std::vector<const Const*>& components)
+      : CompositeConst(ty), component_type_(ty->element_type()) {
+    for (auto c : components) {
+      AppendComponent(c);
+    }
+  }
+
+  VectorConst* AsVectorConst() override { return this; }
+  const VectorConst* AsVectorConst() const override { return this; }
+
+  std::unique_ptr<VectorConst> CopyVectorConst() const {
+    auto another = MakeUnique<VectorConst>(type_->AsVector());
+    another->components_.insert(another->components_.end(), components_.begin(),
+                                components_.end());
+    return another;
+  }
+  std::unique_ptr<Const> Copy() const override {
+    return std::unique_ptr<Const>(CopyVectorConst().release());
+  }
+
+  const analysis::Type* component_type() { return component_type_; }
+
+ private:
+  void AppendComponent(const Const* c) {
+    assert(
+        (c->AsScalarConst() || c->AsNullConst()) &&
+        "The component of vector constant must be of scalar type or null type");
+    assert(c->type() == component_type_ &&
+           "Type mismatch for the component to be appended");
+    components_.push_back(c);
```

Comment:
Actually I just removed this function from the interface. I think the construction of a composite constant should be done at once, nothing should be added once the construction is done, and the number of the components should match with the info stored in the `type_` (so far this is not checked). 

So I think this function should be only used by VectorConstant and place in its private section. Probably we should have a new name for it? It is actually a checker that asserts scalar component constants.